File: scratch/generate_icons.js
const fs = require('fs');
const path = require('path');
const zlib = require('zlib');

// Helper to create uncompressed PNG buffer in pure Node.js
function createPng(width, height, r, g, b) {
  // Signature
  const signature = Buffer.from([137, 80, 78, 71, 13, 10, 26, 10]);

  // IHDR Chunk
  const ihdr = Buffer.alloc(13);
  ihdr.writeUInt32BE(width, 0);
  ihdr.writeUInt32BE(height, 4);
  ihdr[8] = 8; // Bit depth
  ihdr[9] = 2; // Truecolor (RGB)
  ihdr[10] = 0; // Compression
  ihdr[11] = 0; // Filter
  ihdr[12] = 0; // Interlace
  const ihdrChunk = createChunk('IHDR', ihdr);

  // Raw Image Data (with filter byte 0 at start of each scanline)
  const lineLength = width * 3 + 1;
  const rawData = Buffer.alloc(height * lineLength);

  for (let y = 0; y < height; y++) {
    const lineStart = y * lineLength;
    rawData[lineStart] = 0; // None filter
    for (let x = 0; x < width; x++) {
      const idx = lineStart + 1 + x * 3;
      // Draw rounded gradient badge effect
      const dx = x - width / 2;
      const dy = y - height / 2;
      const dist = Math.sqrt(dx * dx + dy * dy);
      if (dist < width * 0.45) {
        // Inner logo color (Golden Yellow #FFB900)
        rawData[idx] = 255;
        rawData[idx + 1] = 185;
        rawData[idx + 2] = 0;
      } else {
        // Background gradient (Teal #008080 to Dark Blue #0F172A)
        rawData[idx] = Math.max(0, Math.min(255, r + Math.floor(x * 10 / width)));
        rawData[idx + 1] = Math.max(0, Math.min(255, g + Math.floor(y * 20 / height)));
        rawData[idx + 2] = b;
      }
    }
  }

  const compressedData = zlib.deflateSync(rawData);
  const idatChunk = createChunk('IDAT', compressedData);

  // IEND Chunk
  const iendChunk = createChunk('IEND', Buffer.alloc(0));

  return Buffer.concat([signature, ihdrChunk, idatChunk, iendChunk]);
}

function createChunk(type, data) {
  const len = data.length;
  const buf = Buffer.alloc(8 + len + 4);
  buf.writeUInt32BE(len, 0);
  buf.write(type, 4, 4, 'ascii');
  data.copy(buf, 8);
  const crc = crc32(buf.subarray(4, 8 + len));
  buf.writeUInt32BE(crc, 8 + len);
  return buf;
}

// CRC32 implementation
function crc32(buf) {
  let crc = 0xffffffff;
  for (let i = 0; i < buf.length; i++) {
    const byte = buf[i];
    crc ^= byte;
    for (let j = 0; j < 8; j++) {
      if (crc & 1) {
        crc = (crc >>> 1) ^ 0xedb88320;
      } else {
        crc = crc >>> 1;
      }
    }
  }
  return (crc ^ 0xffffffff) >>> 0;
}

const assetsDir = path.join(__dirname, '..', 'assets');
if (!fs.existsSync(assetsDir)) {
  fs.mkdirSync(assetsDir, { recursive: true });
}

fs.writeFileSync(path.join(assetsDir, 'icon16.png'), createPng(16, 16, 15, 118, 110));
fs.writeFileSync(path.join(assetsDir, 'icon48.png'), createPng(48, 48, 15, 118, 110));
fs.writeFileSync(path.join(assetsDir, 'icon128.png'), createPng(128, 128, 15, 118, 110));

console.log('Successfully generated PNG icons in assets/!');

File: scratch/generate_icons.py
import zlib
import struct
import math

def clamp(val):
    return max(0, min(255, int(val)))

def make_png(width, height, pixel_func):
    raw_data = bytearray()
    for y in range(height):
        raw_data.append(0) # filter type 0
        for x in range(width):
            r, g, b, a = pixel_func(x, y, width, height)
            raw_data.extend([clamp(r), clamp(g), clamp(b), clamp(a)])
            
    compressed = zlib.compress(raw_data, level=9)
    
    png = bytearray(b'\x89PNG\r\n\x1a\n')
    
    # IHDR
    ihdr = struct.pack('>IIBBBBB', width, height, 8, 6, 0, 0, 0)
    ihdr_crc = zlib.crc32(b'IHDR' + ihdr)
    png.extend(struct.pack('>I', len(ihdr)) + b'IHDR' + ihdr + struct.pack('>I', ihdr_crc))
    
    # IDAT
    idat_crc = zlib.crc32(b'IDAT' + compressed)
    png.extend(struct.pack('>I', len(compressed)) + b'IDAT' + compressed + struct.pack('>I', idat_crc))
    
    # IEND
    iend_crc = zlib.crc32(b'IEND')
    png.extend(struct.pack('>I', 0) + b'IEND' + struct.pack('>I', iend_crc))
    
    return bytes(png)

def render_icon(x, y, w, h):
    nx = (x / (w - 1) if w > 1 else 0.5) * 2 - 1
    ny = (y / (h - 1) if h > 1 else 0.5) * 2 - 1
    dist = math.sqrt(nx * nx + ny * ny)
    
    # Rounded box boundary
    box_d = max(abs(nx), abs(ny))
    if abs(nx) > 0.65 and abs(ny) > 0.65:
        corner_d = math.sqrt((abs(nx)-0.65)**2 + (abs(ny)-0.65)**2)
        if corner_d > 0.33:
            return (0, 0, 0, 0)
            
    # Border cyan glow
    if box_d > 0.86:
        alpha = 255 * (1 - (box_d - 0.86) / 0.14)
        return (0, 242, 254, alpha)
        
    # High-tech Bolt emblem in center
    # Triangle 1 (top-right to center)
    is_bolt = False
    if ny <= 0.05 and ny >= -0.65 and nx >= -0.3 and nx <= 0.35:
        if ny >= -0.65 + ((nx + 0.1) * 1.5):
            is_bolt = True
    if ny >= -0.05 and ny <= 0.65 and nx >= -0.35 and nx <= 0.3:
        if ny <= 0.65 + ((nx - 0.1) * 1.5):
            is_bolt = True

    if is_bolt:
        t = (ny + 0.65) / 1.3
        r = 0 + (168 - 0) * t
        g = 242 + (85 - 242) * t
        b = 254 + (247 - 254) * t
        return (r, g, b, 255)
        
    # Background gradient: #0b1329 to #1e1b4b
    t_bg = (ny + 1) / 2
    bg_r = 11 + (30 - 11) * t_bg
    bg_g = 19 + (27 - 19) * t_bg
    bg_b = 41 + (75 - 41) * t_bg
    
    # Inner subtle radial glow
    glow = max(0.0, 1.0 - dist * 1.1)
    bg_r += 15 * glow
    bg_g += 35 * glow
    bg_b += 65 * glow
    
    return (bg_r, bg_g, bg_b, 255)

for size in [16, 48, 128]:
    data = make_png(size, size, render_icon)
    with open(f'assets/icon{size}.png', 'wb') as f:
        f.write(data)
    print(f'Generated assets/icon{size}.png ({len(data)} bytes)')
